database/dgraph/dgraph.go

```diff
@@ -0,0 +1,512 @@
+	uatomic "go.uber.org/atomic"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/encoding/gzip"
+)
+
+const (
+	unlockedVal = 0
+	lockedVal   = 1
+)
+
+// Driver errors
+var (
+	ErrNilConfig     = errors.New("no config")
+	ErrNoSchema      = errors.New("no schema")
+	ErrDatabaseDirty = errors.New("database is dirty")
+	ErrLockHeld      = errors.New("unable to obtain lock")
+	ErrLockNotHeld   = errors.New("unable to release already released lock")
+)
+
+var DqlJson = jsoniter.Config{TagKey: "dql"}.Froze()
+var GraphQLJson = jsoniter.Config{TagKey: "graphql"}.Froze()
+
+func init() {
+	db := DGraph{}
+	database.Register("dgraph", &db)
+}
+
+const DQLSchema = `
+<version>: int .
+<dirty>: bool .
+type <Migration> {
+	version
+	dirty
+}`
+
+const GraphQLSchema = `
+<Migration.version>: int .
+<Migration.dirty>: bool .
+type <Migration> {
+	Migration.version
+	Migration.dirty
+}`
+
+const GetVersionDQL = `{
+	getVersion(func: type(Migration), first: 1) {
+		version
+		dirty
+	}
+}`
+
+const GetVersionGraphQL = `{
+	getVersion(func: type(Migration), first: 1) {
+		Migration.version
+		Migration.dirty
+	}
+}`
+
+type MutationOperation struct {
+	SetJson    map[string]interface{} `json:"set_json"`
+	DeleteJson map[string]interface{} `json:"delete_json"`
+}
+
+func (m *MutationOperation) ToMutation() (*api.Mutation, error) {
+	mutation := &api.Mutation{}
+
+	if len(m.SetJson) > 0 {
+		setJson, err := json.Marshal(m.SetJson)
+		if err != nil {
+			return nil, err
+		}
+		mutation.SetJson = setJson
+	}
+	if len(m.DeleteJson) > 0 {
+		deleteJson, err := json.Marshal(m.DeleteJson)
+		if err != nil {
+			return nil, err
+		}
+		mutation.DeleteJson = deleteJson
+	}
+	return mutation, nil
+}
+
+type UpsertOperation struct {
+	Query     string               `json:"query"`
+	Mutations []*MutationOperation `json:"mutations"`
+}
+
+type Operation_DropOp string
+
+const (
+	Operation_NONE Operation_DropOp = "NONE"
+	Operation_ALL  Operation_DropOp = "ALL"
+	Operation_DATA Operation_DropOp = "DATA"
+	Operation_ATTR Operation_DropOp = "ATTR"
+	Operation_TYPE Operation_DropOp = "TYPE"
+)
+
+type AlterOperation struct {
+	Schema    []string         `json:"schema"`
+	DropAttr  string           `json:"drop_attr"`
+	DropValue string           `json:"drop_value"`
+	DropOp    Operation_DropOp `json:"drop_op"`
+	DropAll   bool             `json:"drop_all"`
+}
+
+func (a *AlterOperation) ToOperation() *api.Operation {
+	operation := &api.Operation{
+		Schema:    strings.Join(a.Schema, "\n"),
+		DropAttr:  a.DropAttr,
+		DropAll:   a.DropAll,
+		DropValue: a.DropValue,
+	}
+
+	switch a.DropOp {
+	case Operation_NONE:
+		operation.DropOp = api.Operation_NONE
+	case Operation_ALL:
+		operation.DropOp = api.Operation_ALL
+	case Operation_DATA:
+		operation.DropOp = api.Operation_DATA
+	case Operation_ATTR:
+		operation.DropOp = api.Operation_ATTR
+	case Operation_TYPE:
+		operation.DropOp = api.Operation_TYPE
+	}
+	return operation
+}
+
+type Operations struct {
```

Comment:
Is there a usecase for supporting alter, mutate, and upsert in the same migration? If so, the ordering of these operations aren't configurable as is which limits its flexibility. You're better of supporting a `[]Operation` which has a type attribute (alter, mutate, upsert).